server/src/main/java/org/apache/iotdb/db/query/dataset/udf/UDTFFragmentDataSet.java

```
@@ -0,0 +1,122 @@
+
+  private final QueryDataSet fragmentDataSet;
+  private final BlockingQueue<Object[]> productionBlockingQueue;
+
+  private RowRecord[] rowRecords = null;
+  private int rowRecordsLength = 0;
+  private int rowRecordsIndexConsumed = -1;
+
+  private boolean hasNextRowRecords = true;
+
+  public UDTFFragmentDataSet(RawQueryInputLayer rawQueryInputLayer, LayerPointReader[] transformers)
+      throws QueryProcessException, IOException {
+    fragmentDataSet = new UDTFAlignByTimeDataSet(rawQueryInputLayer, transformers);
+    productionBlockingQueue = new LinkedBlockingQueue<>(BLOCKING_QUEUE_CAPACITY);
+    submitTask();
+  }
+
+  @Override
+  public boolean hasNextWithoutConstraint() {
+    try {
+      return rowRecords != null || tryToTakeNextRowRecords();
+    } catch (InterruptedException e) {
+      return onThrowable(e);
+    }
+  }
+
+  private boolean tryToTakeNextRowRecords() throws InterruptedException {
+    if (!hasNextRowRecords) {
+      return false;
+    }
+
+    Object[] production = productionBlockingQueue.take();
+
+    Object rowRecordArrayOrThrowable = production[0];
+    if (rowRecordArrayOrThrowable instanceof Throwable) {
+      return onThrowable((Throwable) rowRecordArrayOrThrowable);
+    }
+
+    rowRecords = (RowRecord[]) production[0];
+    rowRecordsLength = (int) production[1];
+    rowRecordsIndexConsumed = -1;
+    hasNextRowRecords = (boolean) production[2];
+
+    if (rowRecordsLength == 0) {
+      rowRecords = null;
+      // assert !hasNextRowRecords;
+      return false;
+    }
+
+    if (hasNextRowRecords) {
+      submitTask();
+    }
+
+    return true;
+  }
+
+  @Override
+  public RowRecord nextWithoutConstraint() {
+    RowRecord rowRecord = rowRecords[++rowRecordsIndexConsumed];
+    if (rowRecordsIndexConsumed == rowRecordsLength - 1) {
+      rowRecords = null;
+    }
+    return rowRecord;
+  }
+
+  private void submitTask() {
+    if (productionBlockingQueue.remainingCapacity() > 0) {
```

What's the consquence of running a query when `productionBlockingQueue.remainingCapacity() == 0`? It's more likely that the task will be dropped and the next time we call `tryToTakeNextRowRecords()`, it will be blocked in `productionBlockingQueue.take()` forever(`hasNextRoweRecords == true` and the previous call of `tryToTakeNextRowRecords()` is true). I think it's better to throw an exception when the queue is full.